library/core/build.gradle

```diff
@@ -37,6 +37,8 @@ dependencies {
     api project(modulePrefix + 'library-common')
     api project(modulePrefix + 'library-extractor')
     implementation 'androidx.annotation:annotation:' + androidxAnnotationVersion
+    implementation 'androidx.appcompat:appcompat:1.1.0'
```

Comment:
I'm not convinced we want to start taking extra dependencies in core for this. Core currently has exactly one dependency, on a module that provides only annotations. See also the first rule at https://jlbp.dev/ :).

Is there a reason why you can't maintain your own `AudioCapabilitiesReceiver` that contains the workaround in this pull request? I don't think it's actually used inside the library anywhere, so it's possibly only used from your own application code anyway?

Reply:
Sure @ojw28 my [comment](https://github.com/google/ExoPlayer/pull/8178#discussion_r527150635) above speaks to the question of where the code should land up that uses this class. 

The class itself needs to listen to route changes from `MediaRouter` and that means including the AndroidX MediaRouter AAR.   We only see the results of the refactoring of core, so whatever you think is best for this `com.google.android.exoplayer2.audio` package.   class does depend (obviously) on `AudioCapabilities` but that pulls a pretty long thread all the way to the `MediaCodecAudioRenderer`

Reply:
Related to Kevin's earlier comment, if we implemented querying devices using `AudioDeviceInfo` from API 23 would this change still be needed? If not that seems like a better long term solution, and keeping the workaround from this change in the app for now seems preferable to adding this then removing it.

Reply:
Right. In which case, what's the minimum change that _needs_ to go into the library, if anything, to support a workaround in the application?

* I don't think `AudioCapabilitiesReceiver` needs to change at all
* I don't think `TrackSelector.invalidate` needs making public because it's trivial to extend `DefaultTrackSelector` and add a method like `public void forceInvalidate() { invalidate(); }`. That seems preferable, because otherwise we'll be exposing a method in the library API that never does anything useful (unless the app is making other customizations, in which case it's reasonable to expect the app to customize this as well).

The ability to change the `AudioCapabilities` that the `DefaultAudioSink` has seems like the one piece that's missing for an easy application workaround. Is that the case? If so then IMO we should focus on a smaller change that supports replacing those, if it's a feasible thing to do. It would be preferable to do this without removing the immutability property of `AudioCapabilities`.

Reply:
To answer Andrew (@andrewlewis ):  The change to `AudioCapabilitiesReceiver` allows it to detect the change in audio routing, that is essential to the fix.   I've not seen the design for using an `AudioDeviceInfo` query, but it seems to me in any case we will need:

1. some set of `BroadcastReceiver`'s that are notified of the audio route change (HDMI hotplug, Bluetooth device attaches, etc)
2. mechanism to reconfigure the CODEC / Renderer / AudioTrack to work with the new route.

To answer Oliver (@ojw28)  I like the idea of extending `DefaultTrackSelector` and put it in the package with `SimpleExoPlayer` so we can implement option 3 and still keep everything package private at highest access.  I do think it is better to include the fix in the library, otherwise the code the application would need is rather obscure (see the code snippet in my comment above.)  We are putting the listener in `SimpleExoPlayer` ourselves so our applications do not need to include it.

Reply:
I believe the MediaRouter Androidx library is currently too big to be added as an ExoPlayer dependency. If we want to provide the proposed dynamic routing mechanism, we should do so in a MediaRouter ExoPlayer extension that clients can chose to use or not depending on their need and APK size requirements.

Reply:
> Otherwise the code the application would need is rather obscure

This is true, but the problem is also somewhat obscure. It certainly doesn't seem to be widely reported, and it sounds a bit like there's a better long term fix anyway. So I don't really see the justification for these changes needing to go into the library, particularly given they'll either require a (likely unacceptable) additional dependency, or an entirely separate library module.

Reply:
@ojw28 FWIW Issue #7788 occurs on the new Google TV as well when it's connected to any media center (HDMI sound-bar or receiver that supports DD sound), so you may get more reports down the line.   I can't imagine a use case on phones that would run into this, might be one explanation for why you have not seen it reported.

I still think the core library (or some module) will need to listen for and respond to audio routing changes as part of any long term fix.  We can certainly move all the code accept for the change to `AudioCapabilities` to our own code, would be good to have `AudioCapabilitiesReceiver` so we would not need to copy all of that code.

Reply:
Seems like #8318 is related to this?
But its not about bluetooth, just format changes, could your PR help in this case too?